File: rust/xain-grpc/src/bin/numproto_server.rs
use std::sync::Arc;
use std::thread;
use std::time::Duration;

use futures::future::Future;

use xain_grpc::proto::hellonumproto::{NumProtoReply, NumProtoRequest};
use xain_grpc::proto::hellonumproto_grpc::{create_num_proto_server, NumProtoServer};
use xain_grpc::proto::ndarray::NDArray;

type DynError = Box<dyn std::error::Error + Send + Sync>;

#[derive(Clone)]
pub struct NumProtoService;

impl NumProtoServer for NumProtoService {
    fn say_hello_num_proto(
        &mut self,
        ctx: grpcio::RpcContext,
        req: NumProtoRequest,
        sink: grpcio::UnarySink<NumProtoReply>,
    ) {
        // Get the `arr` field from the request.
        let arr: &[u8] = req.get_arr().get_ndarray();
        log::info!("NumProto server received: {:?}", arr);

        // Multiply `arr` by 2.
        let arr: Vec<u8> = arr.iter().map(|x| 2 * x).collect();
        log::info!("NumProto server sent: {:?}", arr);

        // Create a reply with the new value of `arr`.
        let mut nda = NDArray::new();
        nda.set_ndarray(arr);
        let mut reply = NumProtoReply::default();
        reply.set_arr(nda);

        // Spawn a future that sends the reply.
        let fut = sink.success(reply);
        let fut = fut.map_err(|err| log::error!("reply failed: {}", err));
        ctx.spawn(fut)
    }
}

fn main() -> Result<(), DynError> {
    // Set up logging.
    env_logger::Builder::new().filter_level(log::LevelFilter::Info).init();
    grpcio::redirect_log();

    // Create a gRPC event loop.
    let env = Arc::new(grpcio::Environment::new(2));

    // Start a NumProto server.
    let mut server = grpcio::ServerBuilder::new(env)
        .register_service(create_num_proto_server(NumProtoService))
        .bind("127.0.0.1", 50_051)
        .build()?;
    server.start();

    // Sleep forever.
    loop {
        thread::sleep(Duration::from_secs(1_000_000));
    }
}

File: xain_fl/grpc/numproto_server.py
import os
import time
from concurrent import futures

import grpc
from numproto import ndarray_to_proto, proto_to_ndarray

from xain_fl.grpc import hellonumproto_pb2, hellonumproto_pb2_grpc
from xain_fl.logger import get_logger

_ONE_DAY_IN_SECONDS = 60 * 60 * 24
logger = get_logger(__name__, level=os.environ.get("XAIN_LOGLEVEL", "INFO"))


class NumProtoServer(hellonumproto_pb2_grpc.NumProtoServerServicer):
    def SayHelloNumProto(self, request, context):
        nda = proto_to_ndarray(request.arr)
        logger.info("NumProto server received: %s", nda)

        nda *= 2
        logger.info("NumProto server sent: %s", nda)
        return hellonumproto_pb2.NumProtoReply(arr=ndarray_to_proto(nda))


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    hellonumproto_pb2_grpc.add_NumProtoServerServicer_to_server(
        NumProtoServer(), server
    )
    server.add_insecure_port("[::]:50051")
    server.start()
    try:
        while True:
            time.sleep(_ONE_DAY_IN_SECONDS)
    except KeyboardInterrupt:
        server.stop(0)


if __name__ == "__main__":
    serve()
